pkg/operatorstatus/writer.go

```diff
@@ -0,0 +1,101 @@
+package operatorstatus
+
+import (
+	"errors"
+	"reflect"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configclient "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+// statusReporter is used to communicate with the kubernetes api
+// to update the ClusterOperator.
+type writer struct {
+	discovery discovery.DiscoveryInterface
+	client    *configclient.ConfigV1Client
+}
+
+// NewStatusReporter returns a new statusReporter.
+func newWriter() (*writer, error) {
+	cfg, err := config.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := configclient.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	k8sInterface, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &writer{
+		client:    client,
+		discovery: k8sInterface.Discovery(),
+	}, nil
+}
+
+// IsAPIAvailable return true if cluster operator API is present on the cluster.
+// Otherwise, exists is set to false.
+func (w *writer) isAPIAvailable() (exists bool, err error) {
```

Comment:
This should be done outside of this context. The writer should exist only if the API is present.

Reply:
This would again prevent us from recovering from a missing `CO`.